File: System-Design/Circuit-Breaker/cb.ts

enum State { CLOSED, OPEN, HALF_OPEN }

class CircuitBreaker {
    private state: State = State.CLOSED;
    private failureCount: number = 0;
    private readonly failureThreshold: number;
    private readonly cooldownPeriodMs: number;
    private nextAttemptTime: number = 0;

    constructor(threshold: number, cooldownMs: number) {
        this.failureThreshold = threshold;
        this.cooldownPeriodMs = cooldownMs;
    }

    public async execute(action: () => boolean): Promise<string> {
        if (this.state === State.OPEN) {
            if (Date.now() < this.nextAttemptTime) {
                return "FAST_FAIL: Circuit is OPEN";
            }
            this.state = State.HALF_OPEN;
        }

        try {
            const success = action();
            if (!success) throw new Error("Action failed");
            
            this.reset();
            return "SUCCESS";
        } catch (error) {
            this.recordFailure();
            return "FAIL: Action threw an error";
        }
    }

    private recordFailure(): void {
        this.failureCount++;
        if (this.failureCount >= this.failureThreshold) {
            this.state = State.OPEN;
            this.nextAttemptTime = Date.now() + this.cooldownPeriodMs;
        }
    }

    private reset(): void {
        this.failureCount = 0;
        this.state = State.CLOSED;
    }
}

const cb = new CircuitBreaker(2, 100); 
let alwaysFail = () => false;

(async () => {
    await cb.execute(alwaysFail); 
    await cb.execute(alwaysFail); 
    const fastFail = await cb.execute(alwaysFail);
    
    if (fastFail.includes("FAST_FAIL")) {
        console.log("TypeScript Test Passed!");
    } else {
        process.exit(1);
    }
})();

File: System-Design/Circuit-Breaker/cb.cs
using System;

public class Program {
    enum State { CLOSED, OPEN, HALF_OPEN }

    class CircuitBreaker {
        private State state = State.CLOSED;
        private int failureCount = 0;
        private readonly int failureThreshold;
        private readonly long cooldownTicks;
        private long nextAttemptTicks = 0;

        public CircuitBreaker(int threshold, int cooldownMs) {
            failureThreshold = threshold;
            cooldownTicks = cooldownMs * TimeSpan.TicksPerMillisecond;
        }

        public string Execute(Func<bool> action) {
            if (state == State.OPEN) {
                if (DateTime.UtcNow.Ticks < nextAttemptTicks) {
                    return "FAST_FAIL";
                }
                state = State.HALF_OPEN;
            }

            try {
                if (!action()) throw new Exception("Action failed");
                Reset();
                return "SUCCESS";
            } catch {
                RecordFailure();
                return "FAIL";
            }
        }

        private void RecordFailure() {
            failureCount++;
            if (failureCount >= failureThreshold) {
                state = State.OPEN;
                nextAttemptTicks = DateTime.UtcNow.Ticks + cooldownTicks;
            }
        }

        private void Reset() {
            failureCount = 0;
            state = State.CLOSED;
        }
    }
    public static int Main() {
        var cb = new CircuitBreaker(2, 100);
        Func<bool> alwaysFail = () => false;
        
        cb.Execute(alwaysFail);
        cb.Execute(alwaysFail); 
        string res = cb.Execute(alwaysFail); 
        
        if (res == "FAST_FAIL") {
            Console.WriteLine("C# Test Passed!");
            return 0;
        }
        return 1;
    }
}